### PR TITLE
Add USB and Ethernet drivers to support the Raspberry Pi 4

### PR DESCRIPTION
**Pull request by alotipac**
These commits to support the Raspberry Pi 4 add a filter driver that enables the USB 3 controller with RAM above 3 GB and a driver for the Ethernet controller.

### CODE REVIEW

**Comment by worproject** [comment]
Genet works wonderfully. Thanks for the great work!

We've also tested the USB filter driver and it sets the 32-bit accesses flag correctly, but removing the 3 GB limit leads to non-working USB ports on Windows 10 ARM64. Any clues on that?

**Comment by alotipac** [comment]
Try disabling fast startup:

```
"HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Power"
  DWORD HiberbootEnabled = 0
```

**Comment by NiQ1** [comment]
> Genet works wonderfully. Thanks for the great work!
> 
> We've also tested the USB filter driver and it sets the 32-bit accesses flag correctly, but removing the 3 GB limit leads to non-working USB ports on Windows 10 ARM64. Any clues on that?

Seems that if you have patched USBXHCI.SYS it'll break USB (no matter how much RAM) - After installing the driver I needed to revert back to the original USBXHCI.SYS.

**Comment by pbatard** [comment]
> Seems that if you have patched USBXHCI.SYS it'll break USB (no matter how much RAM) - After installing the driver I needed to revert back to the original USBXHCI.SYS.

Yes, that's intentional. The patch values I provided for `winpatch`, that everybody seems to be using, revert the 32 vs 64-bit access logic. I did that so that, if Microsoft ever patched the driver to use 32-bit access on RPi4 with an insider release, people would complain that they couldn't boot any more, and I'd know that something had changed...

So, if you do apply the `HKLM\System\CurrentControlSet\Control\Compatibility\Device\USBXHCI:ACPI!VEN_PNP&DEV_0D10` `USBXHCI` registry key patch, that sets accesses to 32-bit, as per @alotipac's driver (Very nice work figuring that we could set the flag there!), then the patch will invert that logic to force 64-bit access and you'll be back to square one.

In other words, you should not use both a patched `USBXHCI.SYS` and the `USBXHCI` registry flags override. Instead you should only use the registry flags, as it's the much better solution (since it's resilient to a Microsoft driver update and doesn't require breaking the driver signature).

**Comment by pbatard** [comment]
> Try disabling fast startup:
> ```
> "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Power"
>   DWORD HiberbootEnabled = 0
> ```

That was it. 😄 

@alotipac, you are awesome. I don't think we can ever thank you enough for your contributions!!

**Comment by NiQ1** [comment]
> > Try disabling fast startup:
> > ```
> > "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Power"
> >   DWORD HiberbootEnabled = 0
> > ```
> 
> That was it. 😄
> 
> @alotipac, you are awesome. I don't think we can ever thank you enough for your contributions!!

Works even without it. Just need to revert USBXHCI.SYS and remove the 3GB restriction in the BIOS.

**Comment by pbatard** [comment]
> Works even without it. Just need to revert USBXHCI.SYS and remove the 3GB restriction in the BIOS.

Didn't work for me when I tested it. But I did boot Windows with 3 GB first and then removed the limit, so fast startup was active, as opposed to disabling on first install.

**Comment by NiQ1** [comment]
> > Works even without it. Just need to revert USBXHCI.SYS and remove the 3GB restriction in the BIOS.
> 
> Didn't work for me when I tested it. But I did boot Windows with 3 GB first and then removed the limit, so fast startup was active, as opposed to disabling on first install.
> 
> I'm just going to add that this will probably be how most people do it, as the UEFI firmware has to have the 3 GB limit on by default for most OSes. But it shouldn't be too difficult to have the image application utility, or even the driver itself, take care of that registry key...

So did I...
One thing that I noticed is that the Windows 10 splash screen disappears. It appears to hang on the UEFI boot screen (the PI logo) and it takes a couple of mins. The USB indeed stops working for a short while while Windows starts but by the time it reaches the login screen it works again.

**Comment by NiQ1** [comment]
> These commits to support the Raspberry Pi 4 add a filter driver that enables the USB 3 controller with RAM above 3 GB and a driver for the Ethernet controller, which partially resolves #55.

Mind if I post binaries?

**Comment by pbatard** [comment]
I already posted some in the [windows-on-arm discord channe](https://discordapp.com/invite/fqRhc8y). These are drivers that I signed using my Authenticode credentials (so that, even if this isn't enough to get the driver accepted without disabling test signing, we get better accountability and provenance than a self-signed cert) and that also adds the following to the `rpiuxflt.inf` to disable fast boot:

```
HKLM,System\CurrentControlSet\Control\Session Manager\Power,\
    HiberbootEnabled,0x10001,0x00000000
```

I'm attaching these drivers below:
* [rpiuxflt.zip](https://github.com/ms-iot/rpi-iotcore/files/5103486/rpiuxflt.zip)
* [bcmgenet.zip](https://github.com/ms-iot/rpi-iotcore/files/5103492/bcmgenet.zip)

To be completely open, I will also post a link to my rpi-iotcore clone, that contains all the changes I applied to produce the binaries above. I'll edit this post once I have done so.

**Comment by QuimoDrivers** [comment]
> I already posted some in the [windows-on-arm discord channel](https://discordapp.com/invite/fqRhc8y). These are drivers that I signed using my Authenticode credentials (so that, even if this isn't enough to get the driver accepted without disabling test signing, we get better accountability and provenance than a self-signed cert) and that also adds the following to the `rpiuxflt.inf` to disable fast boot:
> 
> 
> 
> ```
> 
> HKLM,System\CurrentControlSet\Control\Session Manager\Power,\
> 
>     HiberbootEnabled,0x10001,0x00000000
> 
> ```
> 
> 
> 
> I'm attaching these drivers below:
> 
> * [rpiuxflt.zip](https://github.com/ms-iot/rpi-iotcore/files/5103486/rpiuxflt.zip)
> 
> * [bcmgenet.zip](https://github.com/ms-iot/rpi-iotcore/files/5103492/bcmgenet.zip)
> 
> 
> 
> To be completely open, I will also post a link to my rpi-iotcore clone, that contains all the changes I applied to produce the binaries above. I'll edit this post once I have done so.
> 
> 
> 
>  

What is the first driver linked used for?

**Comment by NiQ1** [comment]
> > I already posted some in the [windows-on-arm discord channel](https://discordapp.com/invite/fqRhc8y). These are drivers that I signed using my Authenticode credentials (so that, even if this isn't enough to get the driver accepted without disabling test signing, we get better accountability and provenance than a self-signed cert) and that also adds the following to the `rpiuxflt.inf` to disable fast boot:
> > ```
> > 
> > HKLM,System\CurrentControlSet\Control\Session Manager\Power,\
> > 
> >     HiberbootEnabled,0x10001,0x00000000
> > ```
> > 
> > 
> > I'm attaching these drivers below:
> > 
> > * [rpiuxflt.zip](https://github.com/ms-iot/rpi-iotcore/files/5103486/rpiuxflt.zip)
> > * [bcmgenet.zip](https://github.com/ms-iot/rpi-iotcore/files/5103492/bcmgenet.zip)
> > 
> > To be completely open, I will also post a link to my rpi-iotcore clone, that contains all the changes I applied to produce the binaries above. I'll edit this post once I have done so.
> 
> What is the first driver linked used for?

Allows you to use more than 3GB of RAM without breaking USB.

**Comment by worproject** [comment]
What about booting from USB? The filter doesn't seem to work in this case.

**Comment by pbatard** [comment]
Updated the drivers from [my fork](https://github.com/pbatard/rpi-iotcore/tree/rpi4-drivers) with the latest patch and I also took this opportunity to add driver version + date (which should help making sure people can find out what drivers they actually installed).
You can download them from: https://github.com/pbatard/rpi-iotcore/releases/tag/rpi4-drivers-1.1

**Comment by NiQ1** [comment]
@alotipac
The latest version of the driver (multiple bounce buffers) completely broke USB to the point of requiring a complete wipe and reflash. Getting a code 43 error (port reset failed).
Even rolling back doesn't work -
* Installing the old INF back - Failed
* Removing the new version completely and installing the old one - Failed
* Removing every single trace of rpiuxflt from both file system and registry, re-patching USBXHCI.SYS with winpatch and re-enabling 3GB limit - Failed
No way of getting USB back up.
